### ISSUE TITLE
Please point the differences between your boundary loss and  the reference paper 20 in your paper

### ISSUE THREAD

**Issue by jiangx25** [LIVIAETS/boundary-loss#32]
I am sorry about that I do not find obvious differences between your boundary loss and the loss proposed by Karimi et al. [20] in your paper, and i do not find the boundary loss in your code. the loss used in 3D image segmentation is also time consuming

**Comment by HKervadec** [comment]
Hi,

To highlight the difference between the two losses, as written in the paper:
![Screenshot_2021-01-25 Boundary loss for highly unbalanced segmentation - 1812 07032 pdf](https://user-images.githubusercontent.com/4191866/105720570-91223700-5ef1-11eb-8cb4-aacb0066277d.png)

> Unlike our boundary loss, computing $D_S$ cannot be done in a single step before training. The distance needs to be re-computed at each epoch during training, for all the images. It also requires to store the whole volume $\Omega$ in memory, as we cannot compute the distance map for only a subset of $\Omega$. These might be important computational and memory limitations, more so when dealing with large images, as is the case for 3D distance maps.

Implementing their loss becomes much more complex: while our loss is simply a pixel-wise multiplication between the softmax and pre-computed distance map, [20] requires more work (which takes time and slows the training down).
```python
class BoundaryLoss():
    def __init__(self, **kwargs):
        # Self.idc is used to filter out some classes of the target mask. Use fancy indexing
        self.idc: List[int] = kwargs["idc"]
        self.nd: str = kwargs["nd"]
        print(f"Initialized {self.__class__.__name__} with {kwargs}")

    def __call__(self, probs: Tensor, dist_maps: Tensor, _: Tensor, __, ___) -> Tensor:
        assert simplex(probs)
        assert not one_hot(dist_maps)

        pc = probs[:, self.idc, ...].type(torch.float32)
        dc = dist_maps[:, self.idc, ...].type(torch.float32)

        multipled = einsum(f"bk{self.nd},bk{self.nd}->bk{self.nd}", pc, dc)

        loss = multipled.mean()

        return loss

class HausdorffLoss():
    """
    Implementation heavily inspired from https://github.com/JunMa11/SegWithDistMap
    """
    def __init__(self, **kwargs):
        # Self.idc is used to filter out some classes of the target mask. Use fancy indexing
        self.idc: List[int] = kwargs["idc"]
        self.nd: str = kwargs["nd"]
        print(f"Initialized {self.__class__.__name__} with {kwargs}")

    def __call__(self, probs: Tensor, target: Tensor, _: Tensor, __, ___) -> Tensor:
        assert simplex(probs)
        assert simplex(target)
        assert probs.shape == target.shape

        B, K, *xyz = probs.shape  # type: ignore

        pc = cast(Tensor, probs[:, self.idc, ...].type(torch.float32))
        tc = cast(Tensor, target[:, self.idc, ...].type(torch.float32))
        assert pc.shape == tc.shape == (B, len(self.idc), *xyz)

        target_dm_npy: np.ndarray = np.stack([one_hot2hd_dist(tc[b].cpu().detach().numpy())
                                              for b in range(B)], axis=0)
        assert target_dm_npy.shape == tc.shape == pc.shape
        tdm: Tensor = torch.tensor(target_dm_npy, device=probs.device, dtype=torch.float32)

        pred_segmentation: Tensor = probs2one_hot(probs).cpu().detach()
        pred_dm_npy: np.nparray = np.stack([one_hot2hd_dist(pred_segmentation[b, self.idc, ...].numpy())
                                            for b in range(B)], axis=0)
        assert pred_dm_npy.shape == tc.shape == pc.shape
        pdm: Tensor = torch.tensor(pred_dm_npy, device=probs.device, dtype=torch.float32)

        delta = (pc - tc)**2
        dtm = tdm**2 + pdm**2

        multipled = einsum(f"bk{self.nd},bk{self.nd}->bk{self.nd}", delta, dtm)

        loss = multipled.mean()

        return loss
```

> i do not find the boundary loss in your code. the loss used in 3D image segmentation is also time consuming

Indeed the latest version of the code has not been uploaded, I will do that very soon.
